core/src/main/java/fj/data/List.java

```diff
@@ -563,7 +562,57 @@ public final void foreachDoEffect(final Effect1<A> f) {
     return bind(c);
   }
 
-  /**
+    /**
+     * Traverse through the List with a might-fail function.
+     *
+     * @param f The function that might fail.
+     * @return  none if applying f fails to any element of the list or f mapped list in some .
+     */
+    public <B> Option<List<B>> traverseOption(final F<A, Option<B>> f) {
+        F2<A, Option<List<B>>, Option<List<B>>> f2 = (a, as) -> map2Option(List::cons, f.f(a), as);
+        return foldRight(f2, Option.some(List.<B>nil()));
+    }
+
+    private static <A, B, C> Option<List<C>> map2Option(F2<A, List<B>, List<C>> f, final Option<A> opt1, final Option<List<B>> opt2) {
+        return opt1.bind(o1 -> opt2.map(o2 -> f.f(o1, o2)));
+    }
+
+    /**
+     * Traverse through the List with a function that might fail with reporting an error.
+     *
+     * @param f The function that produces Either value.
+     * @return  error in left if applying f fails to any element of the list or f mapped list in right.
+     */
+    public <B, E> Either<E, List<B>> traverseEither(final F<A, Either<E, B>> f) {
+        F2<A, Either<E, List<B>>, Either<E, List<B>>> f2 = (A a, Either<E, List<B>> as) -> List.<B, B, B, E>map2Either(List::cons, f.f(a), as);
+        return foldRight(f2, Either.<E, List<B>>right(List.<B>nil()));
+    }
+
```

Comment:
@mperry yeah I agree unless someone else uses that map2x, there is no point having a separate method. I do remember in my first commit I had a taker for map2Option, but then I moved it to list, doesn't add value. It will make sense to have inline way as you suggested.